arch/risc-v/include/syscall.h

```diff
@@ -391,6 +396,64 @@ static inline uintptr_t sys_call6(unsigned int nbr, uintptr_t parm1,
   return r0;
 }
 
+#ifdef CONFIG_ARCH_USE_S_MODE
+
+/* In S-mode the kernel cannot ecall to itself, because ecall will raise
+ * privileges to M-mode, which is not what we want. Instead, use the syscall
+ * trampoline to do the same.
+ */
+
+static inline uintptr_t ksys_call3(unsigned int nbr, uintptr_t parm1,
```

Comment:
Not yet at least. We need both, even though the kernel does not use the sys_call functions, the declarations / definitions are needed by the PROXY_xxx functions.

I would like to keep it like this, I can rename the functions, but I could not think of a better name.

Reply:
> Not yet at least. We need both, even though the kernel does not use the sys_call functions, the declarations / definitions are needed by the PROXY_xxx functions.
> 
> I would like to keep it like this, I can rename the functions, but I could not think of a better name.

The tool which generate PROXY_xxx is written before you add ksys_call3/4, I don't believe it can predict your change and  
then call these new api.

Reply:
It's like this, the PROXY_xxx needs the original sys_call0-6 functions, because it uses them to create call gates for userspace.

The new functions are named differently (ksys_call), because, well, they have to.

Other option is to move the user syscalls to a different file / location, but this is not what I planned to do.

Reply:
Since nobody call ksys_call3/4, I don't understand why we need keep them. Could you explain more detail?

Reply:
They are called by and needed by e.g. `void up_task_start(main_t taskentry, int argc, char *argv[])`

Originally they called sys_callx(), which would turn into ecall(...), the S-mode flag prevents this, and in this case they call ksys_call() instead.

I don't know of a way to present the sys_call functions to the user and to the kernel with different implementations (the functions have the same name). So this is the reason for the different name (ksys_call).

Reply:
Ok, sorry I miss the caller.

Reply:
Since ksys_call is only used by up_xxx internally, it's better to move it to internal header file(e.g. risc_internal.h or new file).

Reply:
@pussuw after more thinking, it's more better to merge ksys_callx into sys_callx like this:
```
static inline uintptr_t sys_call?(unsigned int nbr, ...)
{
  register long r0 asm("a0") = (long)(nbr);
  ...
#ifdef CONFIG_ARCH_USE_S_MODE
...
#else
   ...
#endif
  return r0;
}
```
So, we don't need modify other place.

Reply:
> The reason I chose the name ksys_call() and made the separate functions is analogous to how for example the memory allocators are implemented now. There are separate malloc/kmalloc functions for example. Both are needed. This is the case with the syscall functions also.

But kmm_malloc will map to malloc in flat build, which make kmm_malloc work in all possible mode(flat, protected and kernel). sys_callx should hand in the same way, which mean that either user or kernel could call these function if the want to invoke the kernel function, its the responsibility of sys_callx to map the implementation to the right action base on the environment.

Reply:
> > The reason I chose the name ksys_call() and made the separate functions is analogous to how for example the memory allocators are implemented now. There are separate malloc/kmalloc functions for example. Both are needed. This is the case with the syscall functions also.
> 
> But kmm_malloc will map to malloc in flat build, which make kmm_malloc work in all possible mode(flat, protected and kernel). sys_callx should hand in the same way, which mean that either user or kernel could call these function if the want to invoke the kernel function, its the responsibility of sys_callx to map the implementation to the right action base on the environment.

Yes, I agree it should be a mapping, the details of the implementation could be hidden in e.g. syscall.h. The problem with the function name sys_callx is that it is directly referenced by the libc functions. So in this case what the function should do depends where the reference comes from, which at least to me makes using the same function name very difficult. 

Unless there is a separate build process for libc and kernel ?